src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs

```diff
@@ -0,0 +1,236 @@
+            /// It will create new <see cref="StateSet"/> for the <see cref="Project"/>.
+            /// This will not have any side effect.
+            /// </summary>
+            public StateSet GetOrCreateStateSet(Project project, DiagnosticAnalyzer analyzer)
+            {
+                var stateSet = _hostStates.GetOrCreateStateSet(project.Language, analyzer);
+                if (stateSet != null)
+                {
+                    return stateSet;
+                }
+
+                return _projectStates.GetOrCreateStateSet(project, analyzer);
+            }
+
+            /// <summary>
+            /// Remove given <see cref="ProjectId"/>.
+            /// </summary>
+            public void RemoveStateSet(ProjectId projectId)
+            {
+                _projectStates.RemoveStateSet(projectId);
+            }
+
+            private void RaiseProjectAnalyzerReferenceChanged(ProjectAnalyzerReferenceChangedEventArgs args)
+            {
+                ProjectAnalyzerReferenceChanged?.Invoke(this, args);
+            }
+
+            private static ImmutableDictionary<DiagnosticAnalyzer, StateSet> CreateAnalyzerMap(
+                    string language, IEnumerable<ImmutableArray<DiagnosticAnalyzer>> analyzerCollection)
+            {
+                var builder = ImmutableDictionary.CreateBuilder<DiagnosticAnalyzer, StateSet>();
+                foreach (var analyzers in analyzerCollection)
+                {
+                    foreach (var analyzer in analyzers)
+                    {
+                        // TODO: 
+                        // #1, all de -duplication should move to HostAnalyzerManager
+                        // #2, not sure whether de-duplicatoin of analyzer itself makes sense. this can only happen
+                        //     if user delibrately put same analyzer twice.
+                        if (builder.ContainsKey(analyzer))
+                        {
+                            continue;
+                        }
+
+                        builder.Add(analyzer, new StateSet(language, analyzer));
+                    }
+                }
+
+                return builder.ToImmutable();
+            }
+
+            [Conditional("DEBUG")]
+            private static void VerifyDiagnosticStates(IEnumerable<StateSet> stateSets)
+            {
+                // Ensure diagnostic state name is indeed unique.
+                for (var i = 0; i < s_stateTypeCount; i++)
+                {
+                    var set = new HashSet<ValueTuple<string, string>>();
+
+                    foreach (var stateSet in stateSets)
+                    {
+                        var state = stateSet.GetState((StateType)i);
+
+                        Contract.Requires(set.Add(ValueTuple.Create(state.Language, state.Name)));
+                    }
+                }
+            }
+        }
+
+        private class ProjectAnalyzerReferenceChangedEventArgs : EventArgs
```

Comment:
Can you please put ProjectAnalyzerReferenceChangedEventArgs and StateSet types in a separate source file? Seem like core part of your change and new state management.